api/grpcserver/transaction_service.go

```diff
@@ -272,3 +280,135 @@ func (s TransactionService) TransactionsStateStream(in *pb.TransactionsStateStre
 		// that the service needs to shut down?
 	}
 }
+
+// StreamResults allows to query historical results and subscribe to live data using the same filter.
+func (s TransactionService) StreamResults(in *pb.TransactionResultsRequest, stream pb.TransactionService_StreamResultsServer) error {
+	var (
+		filter    transactions.ResultsFilter
+		sub       *events.BufferedSubscription[types.TransactionWithResult]
+		err       error
+		persisted types.LayerID
+	)
+	if len(in.Address) > 0 {
+		addr := types.BytesToAddress(in.Address)
+		filter.Address = &addr
+	}
+	if len(in.Id) > 0 {
+		var id types.TransactionID
+		copy(id[:], in.Id)
+		filter.TID = &id
+	}
+	if in.Start > 0 {
+		lid := types.NewLayerID(in.Start)
+		filter.Start = &lid
+	}
+	if in.End > 0 {
+		if in.Watch {
+			return status.Error(codes.InvalidArgument, "watch stream should have an empty End argument")
+		}
+		lid := types.NewLayerID(in.End)
+		filter.End = &lid
+	}
+
+	if in.Watch {
+		sub, err = events.SubscribeMatched(resultsMatcher(filter).match)
+		if err != nil {
+			return status.Error(codes.Internal, err.Error())
+		}
+		defer sub.Close()
+		if err := stream.SendHeader(metadata.MD{}); err != nil {
+			return status.Errorf(codes.Unavailable, "can't send header")
+		}
+	}
+
+	var ierr error
+	err = transactions.IterateResults(s.db, filter, func(rst *types.TransactionWithResult) bool {
+		if rst.Layer.After(persisted) {
+			persisted = rst.Layer
+		}
+		ierr = stream.Send(castResult(rst))
+		return ierr == nil
+	})
+	if err == nil {
+		err = ierr
+	}
+	if err != nil {
+		if errors.Is(err, io.EOF) {
+			return nil
```

Comment:
should check if sub == nil before returning?

Reply:
why? subscription will close in defer if it was initialized
i wanted to avoid printing error in the grpc logs if client went away before consuming whole result, need to doublecheck that it works the way i expect though.

Reply:
i thought if there was no historical data but there is a subscription for future data the subscription will be closed prematurely.